### PR TITLE
[Merged by Bors] - feat: port LinearAlgebra.FreeModule.PID

### PR DESCRIPTION
**Pull request by xroblot**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/FreeModule/PID.lean [hunk 1]
```diff
@@ -20,9 +20,9 @@ equivalently it is an `R`-module linearly equivalent to `ι →₀ R` for some `
 
 This file proves a submodule of a free `R`-module of finite rank is also
 a free `R`-module of finite rank, if `R` is a principal ideal domain (PID),
-i.e. we have instances `[is_domain R] [is_principal_ideal_ring R]`.
+i.e. we have instances `[IsDomain R] [IsPrincipalIdealRing R]`.
 We express "free `R`-module of finite rank" as a module `M` which has a basis
-`b : ι → R`, where `ι` is a `fintype`.
+`b : ι → R`, where `ι` is a `FintypeCat`.
```
Comment (semorrison) [inline Mathlib/LinearAlgebra/FreeModule/PID.lean:25]:
```suggestion
`b : ι → R`, where `ι` is a `Fintype`.
```

File: Mathlib/LinearAlgebra/FreeModule/PID.lean [hunk 2]
```diff
@@ -165,58 +168,58 @@ but must also feed in a basis for `M` using `basis_of_pid` to keep the induction
 theorem Submodule.basis_of_pid_aux [Finite ι] {O : Type _} [AddCommGroup O] [Module R O]
     (M N : Submodule R O) (b'M : Basis ι R M) (N_bot : N ≠ ⊥) (N_le_M : N ≤ M) :
     ∃ y ∈ M,
-      ∃ (a : R)(hay : a • y ∈ N),
+      ∃ (a : R)(_hay : a • y ∈ N),
```
Comment (semorrison) [inline Mathlib/LinearAlgebra/FreeModule/PID.lean:171]:
```suggestion
      ∃ (a : R) (_ : a • y ∈ N),
```
and below.

File: Mathlib/LinearAlgebra/FreeModule/PID.lean [hunk 3]
```diff
@@ -327,19 +331,19 @@ noncomputable def Submodule.basisOfPid {ι : Type _} [Finite ι] (b : Basis ι R
...
   exact Sigma.eq rfl (Basis.eq_of_apply_eq <| finZeroElim)
 #align submodule.basis_of_pid_bot Submodule.basisOfPid_bot
 
 /-- A submodule inside a free `R`-submodule of finite rank is also a free `R`-module of finite rank,
 if `R` is a principal ideal domain.
 
-See also the stronger version `submodule.smith_normal_form_of_le`.
+See also the stronger version `Submodule.smithNormalFormOfLe`.
 -/
 noncomputable def Submodule.basisOfPidOfLe {ι : Type _} [Finite ι] {N O : Submodule R M}
```
Comment (semorrison) [inline Mathlib/LinearAlgebra/FreeModule/PID.lean:344]: I think we've decided to capitalise as `LE` in names like this.

File: Mathlib/LinearAlgebra/FreeModule/PID.lean [hunk 4]
```diff
@@ -431,82 +434,91 @@ section SmithNormal
...
+  induction' O using inductionOnRank with M0 ih
+  exact b
+  intro N N_le_M0
+  obtain ⟨m, b'M⟩ := M0.basisOfPid b
   by_cases N_bot : N = ⊥
   · subst N_bot
     exact ⟨0, m, Nat.zero_le _, b'M, Basis.empty _, finZeroElim, finZeroElim⟩
-  obtain ⟨y, hy, a, hay, M', M'_le_M, N', N'_le_N, N'_le_M', y_ortho, ay_ortho, h⟩ :=
-    Submodule.basis_of_pid_aux M N b'M N_bot N_le_M
+  obtain ⟨y, hy, a, _hay, M', M'_le_M, N', _N'_le_N, N'_le_M', y_ortho, _ay_ortho, h⟩ :=
```
Comment (semorrison) [inline Mathlib/LinearAlgebra/FreeModule/PID.lean:473]:
```suggestion
  obtain ⟨y, hy, a, _, M', M'_le_M, N', _N'_le_N, N'_le_M', y_ortho, _, h⟩ :=
```

File: Mathlib/LinearAlgebra/FreeModule/PID.lean (hunk 3 above)
Comment (xroblot) [reply Mathlib/LinearAlgebra/FreeModule/PID.lean:344]: I changed it and changed also `Submodule.basisOfPidOfLe` in the same way. It felt strange to change `Submodule.basisOfPidOfLeSpan` so I left it like that but I can change it too if you want.

Comment (semorrison) [reply Mathlib/LinearAlgebra/FreeModule/PID.lean]: Yes please.

**Comment by semorrison** [comment]
bors d+

**Pull request closed without merging**